src/base/bittorrent/tracker.cpp

```diff
@@ -271,52 +275,95 @@ void Tracker::processAnnounceRequest()
+            , Log::CRITICAL);
+        throw TrackerError(R"(Invalid "info_hash" parameter)");
+    }
 
     announceReq.infoHash = infoHash;
 
     // 2. peer_id
     const auto peerIdIter = queryParams.find(ANNOUNCE_REQUEST_PEER_ID);
-    if (peerIdIter == queryParams.end())
-        throw TrackerError("Missing \"peer_id\" parameter");
+    if (peerIdIter == queryParams.end()) {
+        LogMsg(tr(R"(Embedded Tracker: Missing "peer_id" parameter)"
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::CRITICAL);
+        throw TrackerError(R"(Missing "peer_id" parameter")");
+    }
 
-    if (peerIdIter->size() > PEER_ID_SIZE)
-        throw TrackerError("Invalid \"peer_id\" parameter");
+    if (peerIdIter->size() > PEER_ID_SIZE) {
+        LogMsg(tr(R"(Embedded Tracker: "peer_id" too big, truncated)"
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::WARNING);   
+    }
 
     announceReq.peer.peerId = *peerIdIter;
+    announceReq.peer.peerId.truncate(20);
 
     // 3. port
     const auto portIter = queryParams.find(ANNOUNCE_REQUEST_PORT);
-    if (portIter == queryParams.end())
-        throw TrackerError("Missing \"port\" parameter");
+    if (portIter == queryParams.end()) {
+        LogMsg(tr(R"(Embedded Tracker: Missing "port" parameter)"
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::CRITICAL);
+        throw TrackerError(R"("Missing "port" parameter")");
+    }
 
     const ushort portNum = portIter->toUShort();
-    if (portNum == 0)
-        throw TrackerError("Invalid \"port\" parameter");
+    if (portNum == 0) {
+        LogMsg(tr(R"(Embedded Tracker: Invalid "port" parameter)"
+                R"( in request from peer: %1, request port: %2)")
+                .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+            , Log::CRITICAL);
+        throw TrackerError(R"("Invalid "port" parameter")");
+    }
 
     announceReq.peer.port = portNum;
 
     // 4. numwant
     const auto numWantIter = queryParams.find(ANNOUNCE_REQUEST_NUM_WANT);
     if (numWantIter != queryParams.end()) {
         const int num = numWantIter->toInt();
-        if (num < 0)
-            throw TrackerError("Invalid \"numwant\" parameter");
-        announceReq.numwant = num;
+        if (num < 0) {
+            LogMsg(tr(R"(Embedded Tracker: Invalid "numwant" parameter, falling back to default)"
+                    R"( in request from peer: %1, request port: %2)")
+                    .arg(announceReq.socketAddress.toString(), announceReq.requestPort)
+                , Log::WARNING);
+        }
+        else { 
+            announceReq.numwant = num;
```

Comment:
:\ This won't do you anything good.

Reply:
Why not? Why is more leniency here not desirable? A warning will still be logged.

Reply:
What if I really want to set it to a negative value and does not want any peer in list? then you would give the unexpected result, no?

Reply:
If negative value isn't acceptable here then the entire request is invalid (unless any fallback is defined by protocol).

Reply:
>If negative value isn't acceptable here then the entire request is invalid (unless any fallback is defined by protocol).

Well, the standard is really obscure on this, IMO this gives room for interpretation. Personally I would treat it as-is and return no peer.

Reply:
@Chocobo1
In that case you just set numwant = 0, same as before. The difference now is that negative values will not result in an error.

@glassez 
https://github.com/qbittorrent/qBittorrent/pull/12434#discussion_r404116723

Reply:
>In that case you just set numwant = 0, same as before. 

Is it really that hard to understand that people/hackers will try to use any value here? You give unexpected results and you just hand waving my concerns? (you do it all the time)
I'm sorry I cannot continue any more. Do as you want, this has nothing to do with me.

Reply:
@Chocobo1

Calm down please, I'm not hand waving anything. The `numwant` field defaults to `50`, so if we don't reassign it, it stays at 50, which is safe.

Previous behavior:

- `num_want > 0`: success.
- `num_want == 0`: success (no peer list returned).
- `num_want < 0`: failure for the announcer and silent failure for the tracker "admin".


New behavior:

- `num_want > 0`: same as before.
- `num_want == 0`: same as before.
- `num_want < 0`: success for the announcer, using our default value for `numwant` of `50`. The tracker "admin" is warned that an announcer has supplied an invalid number, but a default value was used instead to still be able to service the request.

Would you rather I change the text "falling back to default" to "using default value instead" or "using default value of 50 instead" or "using default value of 50"?

Reply:
>If negative value isn't acceptable here then the entire request is invalid (unless any fallback is defined by protocol).

Just to clarify my stance on this, @glassez is right, invalid parameter should stop processing (to avoid wasting more resources) and send back the error message to the peer.

Reply:
The thing is, while the protocol does not define a specific fallback, it does not say the requests should immediately be dropped either.

Reply:
it's not obvious to me that `&numwant=-1` is any more correct (or should be accepted to any greater degree) than `&numwant=foo`.

As far as I can tell (but I haven't thought too much about this either), the only useful values for `numwant` is `0` or a *positive* number.  Clients will generally want as many peers as possible, or 0 (if they are stopping for instance). And the actual number to send has to be up to the tracker to decide.

Reply:
Alright, I removed the leniency changes.